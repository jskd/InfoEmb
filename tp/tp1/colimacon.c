#include <stdlib.h>
#include <stdio.h>
#include <string.h>

/**
 *  Print tableÒ
 */
void print_table(int* table, int cols, int rows)
{
	for(int y=0; y<rows; y++) {
	    for(int x=0; x<cols; x++) {
	    	printf(" %03d ", table[x + y*cols]);
	    }
	    printf("\n");
	}
}

/**
 * Generate table
 * La flemme de factoriser (write right, down, left, top)
 */
void generate_table(int* table, int cols, int rows)
{
	int x=-1; 
	int y=-1; 
	int count=1;
	char write;

    while(1) {

		// write right
	    write=0;

		x++; // go to first case
	    y++; // cancel increment of previous move 

		while(table[x+y*cols] == 0 && x < cols && y < rows) {
			table[x+y*cols]= count; // on ecrit si la case est vide
			count++; 
			x++; // passage a la case suivante
			write= 1;
		}

		if(!write)
			return; // rien ecrit c'est terminé

	    // write down
	    x--;// cancel increment of previous move
	    y++;// go to first case
	    write=0;

		while(table[x+y*cols] == 0 && x < cols && y < rows) {
			table[x+y*cols]= count;
			count++;
			y++;
		    write= 1;
		}

		if(!write)
			return;

	    // write left
	    x--; // go to first case
	    y--; // cancel increment of previous move
	    write=0;

		while(table[x+y*cols] == 0 && x < cols && y < rows) {
			table[x+y*cols]= count;
			count++;
			x--;
		    write= 1;
		}

		if(!write)
			return;

	    // write left

        
	    x++; // cancel increment of previous move
	    y--; // go to first case

	    write=0;

		while(table[x+y*cols] == 0 && x < cols && y < rows) {
			table[x+y*cols]= count;
			count++;
			y--;
			write= 1;
		}

		if(!write)
			return;	    

    }

}

// main
int main(int argc, char** argv)
{
  if(argc < 3)
  {
    printf("Usage: colimacon [cols] [rows]\n");
    return -1;
  }
  int cols= atoi(argv[1]);
  int rows= atoi(argv[2]);

  int table[cols][rows];
  memset (table, 0, sizeof (table));


  generate_table( (int *)table, rows, cols);
  print_table( (int *)table, rows, cols);
}