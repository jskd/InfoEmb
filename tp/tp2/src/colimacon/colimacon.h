/**
 * @file   src/colimacon/colimacon.h
 * @author Jérôme SKODA <contact@jeromeskoda.fr>
 * @date   Tue 17 Oct 2017 01:46:29 PM CEST
 * @brief  Colimaçon header
 */

/**
 * @brief Allocation d'un tableau et remplissage en colimacon
 * @param[in] line nombre de ligne
 * @param[in] col nombre de colonne
 * @return retourne un pointer vers le tableau
 */
int ** colimacon(int line, int col);

/**
 *  @brief Affiche un tableau en colimaçon
 *  @param[in] table table à imprimmer
 *  @param[in] line nombre de ligne de la table
 *  @param[in] col nombre de colonnes de la table
 */
void print_colimacon(int** table, int line, int col);

/**
 *  @brief Supression d'un tableau
 *  @param[in] table table à supprimer
 *  @param[in] line nombre de ligne de la table
 *  @param[in] col nombre de colonnes de la table
 */
void delete_colimacon(int** table, int line, int col);
