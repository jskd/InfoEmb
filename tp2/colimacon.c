#include <stdlib.h>
#include <stdio.h>
#include <string.h>

struct cursor {
  int x;
  int y;
};


/**
 *  Print tableÒ
 */
void print_table(int** table, int cols, int rows)
{
  for(int y=0; y<rows; y++) {
      for(int x=0; x<cols; x++) {
        printf(" %03d ", table[x][y]);
      }
      printf("\n");
  }
}

/**
 * Generate table
 * La flemme de factoriser (write right, down, left, top)
 */
void generate_table(int** table, int cols, int rows)
{
  struct cursor pos= {0,0};

  int x=-1;
  int y=-1;
  int count=1;
  char write;

    while(1) {

    // write right
      write=0;

    x++; // go to first case
    y++; // cancel increment of previous move

    while(table[x][y] == 0 && x < cols && y < rows) {
      table[x][y]= count; // on ecrit si la case est vide
      count++;
      x++; // passage a la case suivante
      write= 1;
    }

    if(!write)
      return; // rien ecrit c'est terminé

      // write down
      x--;// cancel increment of previous move
      y++;// go to first case
      write=0;

    while(table[x][y] == 0 && x < cols && y < rows) {
      table[x][y]= count;
      count++;
      y++;
      write= 1;
    }

    if(!write)
      return;

      // write left
      x--; // go to first case
      y--; // cancel increment of previous move
      write=0;

    while(table[x][y] == 0 && x < cols && y < rows) {
      table[x][y]= count;
      count++;
      x--;
      write= 1;
    }

    if(!write)
      return;

      // write left

      x++; // cancel increment of previous move
      y--; // go to first case

      write=0;

    while(table[x][y] == 0 && x < cols && y < rows) {
      table[x][y]= count;
      count++;
      y--;
      write= 1;
    }

    if(!write)
      return;
    }

}

// main
int main(int argc, char** argv)
{
  if(argc < 3)
  {
    printf("Utilisation: colimacon [colonnes] [lignes]\n");
    return -1;
  }

  int cols= atoi(argv[1]);
  int rows= atoi(argv[2]);
  cols++;
  rows++;

  if(cols < 0 || rows < 0)
  {
    printf("Erreur les colonnes et les lignes doivent être positif.\n");
    return -1;
  }

  int** table= (int**) malloc(cols*sizeof(int *));

  for(int i=0; i<cols; i++) {
    table[i]= (int*) malloc(rows*sizeof(int));
    memset(table[i], 0, sizeof(rows*sizeof(int)));
  }



  generate_table( table, rows, cols);
  print_table( table, rows, cols);

  for(int i=0; i<cols; i++)
    free(table[i]);
  free(table);
}
