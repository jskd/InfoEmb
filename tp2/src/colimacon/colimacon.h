/**
 * @brief Allocation d'un tableau et remplissage en colimacon
 * @param[in] line nombre de ligne
 * @param[in] col nombre de colonne
 * @return retourne un pointer vers le tableau
 */
int ** colimacon(int line, int col);

/**
 *  @brief Affiche un tableau en colimaçon
 *  @param[in] table table à imprimmer
 *  @param[in] line nombre de ligne de la table
 *  @param[in] col nombre de colonnes de la table
 */
void print_colimacon(int** table, int line, int col);

/**
 *  @brief Supression d'un tableau
 *  @param[in] table table à supprimer
 *  @param[in] line nombre de ligne de la table
 *  @param[in] col nombre de colonnes de la table
 */
void delete_colimacon(int** table, int line, int col);
