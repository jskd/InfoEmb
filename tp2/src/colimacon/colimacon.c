#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include "colimacon.h"

struct borne {
  int haute;
  int droite;
  int basse;
  int gauche;
};

struct curseur {
  int col;
  int ligne;
};

typedef enum {
  HAUT,
  GAUCHE,
  BAS,
  DROITE
} direction_t ;

/**
 *  Print table√í
 */
void print_colimacon(int** table, int line, int col)
{
  for(int i_line=0; i_line < line; i_line++) {
      for(int i_col=0; i_col < col; i_col++) {
        printf(" %03d ", table[i_line][i_col] );
      }
      printf("\n");
  }
}

void delete_colimacon(int** table, int line, int col)
{
  for(int i_line; i_line<line; i_line++)
    free(table[i_line]);
  free(table);
}

static int** _malloc_colimacon(int line, int col) {

  int** table= malloc( line * sizeof(int *) );
  for(int i_line=0; i_line < line; i_line++)
    table[i_line] = malloc( col * sizeof(int));

  return table;
}

static void _iter_curseur( struct curseur* c, direction_t d ) {
  switch(d) {
    case HAUT:
      c->ligne--;
      break;
    case BAS:
      c->ligne++;
      break;
    case DROITE:
      c->col++;
      break;
    case GAUCHE:
      c->col--;
      break;
  }
}

static void _iter_borne( struct borne* borne, direction_t d) {
  switch(d) {
    case HAUT:
      borne->gauche++;
      break;
    case BAS:
      borne->droite--;
      break;
    case DROITE:
      borne->haute++;
      break;
    case GAUCHE:
      borne->basse--;
      break;
  }
}

static struct curseur _get_position_initiale(struct borne* borne, direction_t d) {
  struct curseur cur;
  switch(d) {
    case HAUT:
      cur.col = borne->gauche+1;
      cur.ligne = borne->basse-1;
      break;
    case BAS:
      cur.col =  borne->droite-1;
      cur.ligne = borne->haute+1;
      break;
    case GAUCHE:
      cur.col = borne->droite-1;
      cur.ligne = borne->basse-1;
      break;
    case DROITE:
      cur.col = borne->gauche+1;
      cur.ligne = borne->haute+1;
      break;
  }
  return cur;
}

static char _canWrite(struct curseur* cur, struct borne* bor) {
  return cur->col > bor->gauche && cur->col < bor->droite
    && cur->ligne > bor->haute && cur->ligne < bor->basse;
}


static int _ecriture_direction(int** tab, struct borne* bor, int* val,
  direction_t dir)
{
  struct curseur cur= _get_position_initiale(bor, dir);

  if(!_canWrite(&cur, bor)) {
      return -1;
  }

  do {
    tab[cur.ligne][cur.col]= *val;
    (*val)++;
    _iter_curseur(&cur, dir);
  } while(_canWrite(&cur, bor));

  _iter_borne(bor, dir);

  return 0;
}


int** colimacon(int line, int col) {

  struct borne borne;
  borne.gauche= -1;
  borne.haute= -1;
  borne.droite= col;
  borne.basse= line;
  int valeur=1;

  // initlisation du tableau
  int** table= _malloc_colimacon(line, col);

  while(1) {
    if(_ecriture_direction(table, &borne, &valeur, DROITE))
      break;
    if(_ecriture_direction(table, &borne, &valeur, BAS))
      break;
    if(_ecriture_direction(table, &borne, &valeur, GAUCHE))
      break;
    if(_ecriture_direction(table, &borne, &valeur, HAUT))
      break;
  }

  return table;
}

