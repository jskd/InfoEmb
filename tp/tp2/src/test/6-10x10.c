/**
 * @file   src/test/6-10x10.c
 * @author Jérôme SKODA <contact@jeromeskoda.fr>
 * @date   Tue 17 Oct 2017 01:46:29 PM CEST
 * @brief  Test 6
 */

#include "../colimacon/colimacon.h"
#include <stdio.h>
#include <stdlib.h>

static const int _line= 10;
static const int _col= 10;
/**
 * @brief Programme de test
 *        Taille 10x10
 */
int main(int argc, char** argv)
{

  int** table= colimacon(_line, _col);

  print_colimacon(table, _line, _col);

  delete_colimacon(table, _line, _col);

  return 0;
}
