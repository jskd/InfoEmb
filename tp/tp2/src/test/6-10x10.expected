 001  002  003  004  005  006  007  008  009  010 
 036  037  038  039  040  041  042  043  044  011 
 035  064  065  066  067  068  069  070  045  012 
 034  063  084  085  086  087  088  071  046  013 
 033  062  083  096  097  098  089  072  047  014 
 032  061  082  095  100  099  090  073  048  015 
 031  060  081  094  093  092  091  074  049  016 
 030  059  080  079  078  077  076  075  050  017 
 029  058  057  056  055  054  053  052  051  018 
 028  027  026  025  024  023  022  021  020  019 
